Fix `Camera3D` projection issues with `PROJECTION_FRUSTUM`, viewport 2d override and `KEEP_WIDTH`

Flarkk opened the pull request:
Resolve  #77906
Resolve #90074
Resolve #61174

Partially address the limitations mentioned in https://github.com/godotengine/godot-proposals/issues/8536

Depends on #99961 (more unit tests)

Please test it extensively.
I wrote quite a lot of unit tests but it's error prone mathy stuff so the more testers the better.

Flarkk committed:
Add tests for `Camera3D` frustum functions

Flarkk committed:
Fix `Camera3D` projection issues with frustum and viewport 2d override

Limeth commented:
Hello! I just ran into the issue this PR is trying to fix, and was delighted to see your efforts.
I've tried running my project with these changes, and it mostly seems to work fine, except when `Camera3D::keep_aspect` is set to `Camera3D.KeepAspect.KEEP_WIDTH` rather than the default `KEEP_HEIGHT`, then the math seems wrong. Looking at the changes, it is not yet clear to me why.

Flarkk commented:
@Limeth which function is affected ?

Limeth commented:
> @Limeth which function is affected ?

Forgive my unfamiliarity with Godot, but I am seeing the issue on `_mouse_enter()` and `_mouse_exit()` of an `Area3D` node.
I suppose it must be using `project_local_ray_normal` to ray-trace the scene and trigger the event listeners above.

Limeth commented:
A demo with a demonstration.

Correct behavior with `KEEP_HEIGHT`:
https://github.com/user-attachments/assets/dd767c20-3a08-4777-950e-a6b91838923b

Incorrect behavior with `KEEP_WIDTH`:
https://github.com/user-attachments/assets/00f33176-a3c2-446c-97f9-d4beaa860609

Flarkk commented:
Just fixed a couple of issues with the aforementioned functions not taking into account `KEEP_WIDTH` cameras.
More importantly I reworked the test to facilitate programmatical testing of various combinations of `Camera3D.KeepAspect`, `Viewport.size`, `Viewport.size_2d_override`.
This should help catching more issues.
@Limeth it would be great if you could test again.

Limeth commented:
I tested my project with the changes, and unfortunately it seems that neither of the keep_aspect options are working for me now.
Since my project is pretty small so far, I thought it'd be best for me to share it so you can investigate what the issue might be yourself.

[project.zip](https://github.com/user-attachments/files/18101414/project.zip)

You can do the same test as in the videos above just by running the project and holding the right mouse button.
It seems to work, as long as the window isn't resized. Resizing the window changes the keep_aspect option, depending on whether the aspect ratio increased or decreased compared to the original aspect ratio, to always keep the scene within the window.
Resizing also seems to break the ray tracing math.

Let me know if there's anything more I could do to help out.

Flarkk commented:
Bad news... But thanks for taking time. It helps. Will take a look again

Flarkk commented:
Here we are ! I made a small mistake both in the code and in the tests.
I could test it extensively with both your project and the test scene from #61174, with various combinations of `KEEP_HEIGHT`/`KEEP_WIDTH`, viewport size and viewport size override. 
I think I finally got it right.

All this made me realize that `SubViewport.size_2d_override_stretch` doesn't work [as documented](https://docs.godotengine.org/en/stable/classes/class_subviewport.html#class-subviewport-property-size-2d-override-stretch). It's rather a global flag that toggles on / off the whole 2d size overriding logic - not only the stretch aspect of it.

In other words, when `size_2d_override_stretch` is `false`, `size_2d_override` is ignored everywhere.
Note that the base class' `Viewport.size_2d_override_stretch()` always return true and `Window` inherits this behavior. In For those 2 classes `size_2d_override` will be effective right away.

We might want to clarify the documentation at some point.

Limeth commented:
I can confirm that it works properly for me now. Thank you!

Flarkk commented:
@lyuma would you have time to take a look at this ? Cc @fire

l0g0sys commented:
@Flarkk Is there programmatic workaround for this? I need to use unproject_position() on camera with Frustrum projection and KEEP_WIDTH mode.

Flarkk commented:
I think it's doable to reimplement `unproject_position()` in GDScript, and call from there the fixed version of `_get_camera_projection()` in GDScript too :

```c++
Point2 Camera3D::unproject_position(const Vector3 &p_pos) const {
	ERR_FAIL_COND_V_MSG(!is_inside_tree(), Vector2(), "Camera is not inside scene.");

	Size2 viewport_size = get_viewport()->get_visible_rect().size;

	Projection cm = _get_camera_projection(_near); // This must use the fixed version

	Plane p(get_camera_transform().xform_inv(p_pos), 1.0);

	p = cm.xform4(p);

	// Prevent divide by zero.
	// TODO: Investigate, this was causing NaNs.
	ERR_FAIL_COND_V(p.d == 0, Point2());

	p.normal /= p.d;

	Point2 res;
	res.x = (p.normal.x * 0.5 + 0.5) * viewport_size.x;
	res.y = (-p.normal.y * 0.5 + 0.5) * viewport_size.y;

	return res;
}

```

```c++
Projection Camera3D::_get_camera_projection(real_t p_near) const {
	Size2 viewport_size = get_viewport()->get_camera_rect_size();
	Projection cm;

	switch (mode) {
		case PROJECTION_PERSPECTIVE: {
			cm.set_perspective(fov, viewport_size.aspect(), p_near, _far, keep_aspect == KEEP_WIDTH);
		} break;
		case PROJECTION_ORTHOGONAL: {
			cm.set_orthogonal(size, viewport_size.aspect(), p_near, _far, keep_aspect == KEEP_WIDTH);
		} break;
		case PROJECTION_FRUSTUM: {
			cm.set_frustum(size, viewport_size.aspect(), frustum_offset, p_near, _far, keep_aspect == KEEP_WIDTH);
		} break;
	}

	return cm;
}
```

l0g0sys commented:
Yep, looks like lot of fun.
Already re-implemented original unproject_position() along with unexposed C++ methods like xform_inv() and xform4(), but I had to make mistake somewhere as it doesn't return same coords as original one.

Flarkk commented:
If you can share your code I may be able to help.

l0g0sys commented:
So, this is original Camera3D.UnprojectPosition():
```
	private static Vector2 UnprojectPosition(Camera3D camera, Vector3 p_pos)
	{
		Vector2 viewport_size = camera.GetViewport().GetVisibleRect().Size;

		Projection cm = camera.GetCameraProjection();

		Plane p = new Plane(XformInv(camera.Transform, p_pos), 1.0f);

		p = Xform4(cm, p);
		p.Normal /= p.D;

		return new Vector2(
			(p.Normal.X * 0.5f + 0.5f) * viewport_size.X,
			(-p.Normal.Y * 0.5f + 0.5f) * viewport_size.Y
		);
	}
```
Internal C++ Vector3 Transform3D::xform_inv(const Vector3 &p_vector):
```
	private static Vector3 XformInv(Transform3D transform, Vector3 p_vector)
	{
		Vector3 v = p_vector - transform.Origin;

		Basis basis = transform.Basis;

		return new Vector3(
			(basis.Row0[0] * v.X) + (basis.Row1[0] * v.Y) + (basis.Row2[0] * v.Z),
			(basis.Row0[1] * v.X) + (basis.Row1[1] * v.Y) + (basis.Row2[1] * v.Z),
			(basis.Row0[2] * v.X) + (basis.Row1[2] * v.Y) + (basis.Row2[2] * v.Z)
		);
	}

```
And internal C++ Plane Projection::xform4(const Plane &p_vec4):
```
	private static Plane Xform4(Projection projection, Plane p_vec4)
	{
		Vector3 normal = new Vector3(
			projection[0][0] * p_vec4.Normal.X + projection[1][0] * p_vec4.Normal.Y + projection[2][0] * p_vec4.Normal.Z + projection[3][0] * p_vec4.D,
			projection[0][1] * p_vec4.Normal.X + projection[1][1] * p_vec4.Normal.Y + projection[2][1] * p_vec4.Normal.Z + projection[3][1] * p_vec4.D,
			projection[0][2] * p_vec4.Normal.X + projection[1][2] * p_vec4.Normal.Y + projection[2][2] * p_vec4.Normal.Z + projection[3][2] * p_vec4.D
		);

		float d = projection[0][3] * p_vec4.Normal.X + projection[1][3] * p_vec4.Normal.Y + projection[2][3] * p_vec4.Normal.Z + projection[3][3] * p_vec4.D;

		return new Plane(normal, d);
	}
```

Flarkk commented:
> `Projection cm = camera.GetCameraProjection();`

You need to reimplement `GetCameraProjection()` too. This is what this PR ultimately fixes.

l0g0sys commented:
Yeah, i will. But first I have to reimplement old code so I can verify it against real UnprojectPosition method. After I get it working 1:1, then I can apply fix.

l0g0sys commented:
Oh, found the error.
 On line `Plane p = new Plane(XformInv(camera.Transform, p_pos), 1.0f);`
value of `camera.Transform` is different than value of C++ `get_camera_transform()` method.

Flarkk commented:
If you can share your final workaround code in this thread it may be found useful by other users until this PR gets merged.

l0g0sys commented:
Updated those reimplementations above to include also workaround. Didn't want to spam extra set of code.

Flarkk commented:
Rebased using `Math` namespace for `SQRT3` math constant.

Flarkk commented:
Unit tests this PR adds that cover `FRUSTUM` camera mode now fail since this rebase.

I couldn't spend time on it yet. There are 3 possible causes:

1. Precision issues in the tests introduced by `Math::SQRT3` (which is a `constexpr double` unlike the previous local `#define SQRT3 1.732050f`)

2. A mistake a could have made doing the rebase that could have changed the tests' logic

3. A regression affecting projections in `FRUSTUM` camera mode that were introduced by another PR in the meantime

I will switch this PR to draft until it's resolved.

Flarkk commented:
Now tests pass.
It was case no. 3 with `Projection::get_viewport_half_extents()` having a different behavior with non-symmetrical viewports since #100209.

There is now another unrelated compilation error with `Vector2` multiplication I didn't experienced locally.
I should be able to fix it shortly.

Flarkk commented:
Here we are.
Reviews welcome !

Flarkk commented:
Thanks for reporting. If you can somehow provide a MRP, that would help. Will wrap my mind around what you describe anyway.

Flarkk commented:
In your video it looks like that it acts as if the viewport had twice smaller extent from the top-left corner. For instance when the label should be at the center, it's actually positioned on the bottom-right corner. Is this what you observed ?

Flarkk commented:
@benbot could you narrow it down or make a MRP ?

Flarkk commented:
I tried to reproduce what you describe but haven't been very successful so far.
I got somewhat similar behavior with the following scene setup, when `stretch` is **not** enabled on the `ViewportContainer`. Ticking `stretch` solves the problem. Does your scene have a similar setup ?
![image](https://github.com/user-attachments/assets/ff766048-604d-416d-a33e-10f787622b0b)

elvisish commented:
Is it possible to get a fix for 3.6 or 3.7? I'm having to do the calculations with magic numbers:
```
(cam.project_ray_origin(viewport.size * 0.5) + cam.project_ray_normal(viewport.size * 0.5 + Vector2(0,cam.get_frustum_offset().y * -20000)) * ray_length)
```
but unfortunately this breaks if I change the viewport size or size of the frustum camera.

Flarkk commented:
> Is it possible to get a fix for 3.6 or 3.7?

I don't see a reason it cannot be cherry-picked in 3.x.

If it turns out to require a custom implementation I can contribute.

elvisish commented:
> > Is it possible to get a fix for 3.6 or 3.7?
> 
> I don't see a reason this PR cannot be cherry-picked in 3.x.
> 
> If it turns out to require a custom implementation I can contribute.

Thankyou, I think @lawnjelly is the sole maintainer for 3.x, I'm trying to finish up a game I've been working on for 3 years now but after reworking the viewport setup it's broken the hack I was using for this.

Can it be implemented in gdscript currently so I can get frustum offset ray projection as close as possible until it might potentially be backported? I'm thinking about trying some kind of weird hacky rotation gimbal temporarily to get a close enough position, but I'd much rather just try a new projection function that actually takes the offset into account.

elvisish commented:
> > Can it be implemented in gdscript currently so I can get frustum offset ray projection as close as possible until it might potentially be backported?
> 
> Yes it can be worked around in GDScript. See my [above post](https://github.com/godotengine/godot/pull/100121#issuecomment-2763233784).

I saw that, but it kinda looks more like cpp to me? Also, I'm using project_ray_origin and project_ray_normal to test a ray forwards from the camera in the direction of the frustum offset, would those functions be much different reimplemented?

elvisish commented:
> > Can it be implemented in gdscript currently so I can get frustum offset ray projection as close as possible until it might potentially be backported?
> 
> Yes it can be worked around in GDScript. See my [above post](https://github.com/godotengine/godot/pull/100121#issuecomment-2763233784).

I've tried to port this to GDScript, but it's not quite working:
```swift
func unproject_position(p_pos):
	print(!is_inside_tree(), Vector2(), "Camera is not inside scene.")

	var viewport_size : Vector2 = get_viewport().get_visible_rect().size

	var cm = _get_camera_projection(_near) ## This must use the fixed version

	p(get_camera_transform().xform_inv(p_pos), 1.0)

	p = cm.xform4(p);

	## Prevent divide by zero.
	## TODO: Investigate, this was causing NaNs.
	ERR_FAIL_COND_V(p.d == 0, Point2());

	p.normal /= p.d;

	var res
	res.x = (p.normal.x * 0.5 + 0.5) * viewport_size.x
	res.y = (-p.normal.y * 0.5 + 0.5) * viewport_size.y

	return res

func _get_camera_projection(real_t, p_near):
	var viewport_size : Rect2 = get_viewport().get_camera_rect_size()
	var cm

	match (mode):
		PROJECTION_PERSPECTIVE: 
			cm.set_perspective(fov, viewport_size.aspect(), p_near, _far, keep_aspect == KEEP_WIDTH)
		PROJECTION_ORTHOGONAL: 
			cm.set_orthogonal(size, viewport_size.aspect(), p_near, _far, keep_aspect == KEEP_WIDTH)
		PROJECTION_FRUSTUM: 
			cm.set_frustum(size, viewport_size.aspect(), frustum_offset, p_near, _far, keep_aspect == KEEP_WIDTH)

	return cm
```
I'm kinda desperate at this point, but I still need `project_ray_origin` and `project_ray_normal` to be ported and I'm not sure where to start with those.

Flarkk commented:
> I still need project_ray_origin and project_ray_normal to be ported and I'm not sure where to start with those.

As far as I remember `project_ray_origin` is not modified by this PR so you should be able to use the vanilla version.

For `project_ray_normal` you need to rewrite the following in GDScript : 

```c++
Vector3 Camera3D::project_ray_normal(const Point2 &p_pos) const {
	Vector3 ray = project_local_ray_normal(p_pos);
	return get_camera_transform().basis.xform(ray).normalized();
}

Vector3 Camera3D::project_local_ray_normal(const Point2 &p_pos) const {
	ERR_FAIL_COND_V_MSG(!is_inside_tree(), Vector3(), "Camera is not inside scene.");

	Size2 viewport_size = get_viewport()->get_camera_rect_size();
	Vector2 cpos = get_viewport()->get_camera_coords(p_pos);
	Vector3 ray;

	if (mode == PROJECTION_ORTHOGONAL) {
		ray = Vector3(0, 0, -1);
	} else {
		Projection cm = _get_camera_projection(_near);
		Vector2 top_right = cm.get_viewport_half_extents();
		Vector2 offset = Vector2();
		if (mode == PROJECTION_FRUSTUM) {
			offset = frustum_offset;
		}
		ray = Vector3(((cpos.x / viewport_size.width) * 2.0 - 1.0) * top_right.x + offset.x, ((1.0 - (cpos.y / viewport_size.height)) * 2.0 - 1.0) * top_right.y + offset.y, -_near).normalized();
	}

	return ray;
}
```

elvisish commented:
> > I still need project_ray_origin and project_ray_normal to be ported and I'm not sure where to start with those.
> 
> As far as I remember `project_ray_origin` is not affected by this PR because it always returns the camera origin in Frustum and Perspective modes. You should be able to use the vanilla version here.
> 
> For `project_ray_normal` you need to rewrite the following in GDScript :
> 
> ```c++
> Vector3 Camera3D::project_ray_normal(const Point2 &p_pos) const {
> 	Vector3 ray = project_local_ray_normal(p_pos);
> 	return get_camera_transform().basis.xform(ray).normalized();
> }
> 
> Vector3 Camera3D::project_local_ray_normal(const Point2 &p_pos) const {
> 	ERR_FAIL_COND_V_MSG(!is_inside_tree(), Vector3(), "Camera is not inside scene.");
> 
> 	Size2 viewport_size = get_viewport()->get_camera_rect_size();
> 	Vector2 cpos = get_viewport()->get_camera_coords(p_pos);
> 	Vector3 ray;
> 
> 	if (mode == PROJECTION_ORTHOGONAL) {
> 		ray = Vector3(0, 0, -1);
> 	} else {
> 		Projection cm = _get_camera_projection(_near); // This must be the fixed version 
> 		Vector2 top_right = cm.get_viewport_half_extents();
> 		Vector2 offset = Vector2();
> 		if (mode == PROJECTION_FRUSTUM) {
> 			offset = frustum_offset;
> 		}
> 		ray = Vector3(((cpos.x / viewport_size.width) * 2.0 - 1.0) * top_right.x + offset.x, ((1.0 - (cpos.y / viewport_size.height)) * 2.0 - 1.0) * top_right.y + offset.y, -_near).normalized();
> 	}
> 
> 	return ray;
> }
> ```

I've been struggling with this, here's what I have so far:
```swift
extends Camera

func project_ray_normal(p_pos):
	var ray : Vector3 = project_local_ray_normal(p_pos)
	return get_camera_transform().basis.xform(ray).normalized()


func project_local_ray_normal(p_pos):
	print(!is_inside_tree(), Vector3(), "Camera is not inside scene.")
	var viewport_size : Rect2 = get_viewport().get_visible_rect()
	var cpos : Vector2 = get_viewport().get_camera_coords(p_pos)
	var ray : Vector3

	if (projection == PROJECTION_ORTHOGONAL):
		ray = Vector3(0, 0, -1)
	else:
		var cm = _get_camera_projection(near) ## This must be the fixed version 
		var top_right : Vector2 = cm.get_viewport_half_extents()
		var offset : Vector2
		if (projection == PROJECTION_FRUSTUM):
			offset = frustum_offset
		ray = Vector3(((cpos.x / viewport_size.size.x) * 2.0 - 1.0) * top_right.x + offset.x, ((1.0 - (cpos.y / viewport_size.size.y)) * 2.0 - 1.0) * top_right.y + offset.y, -near).normalized()
	return ray

func unproject_position(p_pos):
	print(!is_inside_tree(), Vector2(), "Camera is not inside scene.")

	var viewport_size : Vector2 = get_viewport().get_visible_rect().size

	var cm = _get_camera_projection(near) ## This must use the fixed version

	var p(get_camera_transform().xform_inv(p_pos), 1.0)

	p = cm.xform4(p);

	## Prevent divide by zero.
	## TODO: Investigate, this was causing NaNs.
	print(p.d == 0, Point2())

	p.normal /= p.d;

	var res
	res.x = (p.normal.x * 0.5 + 0.5) * viewport_size.x
	res.y = (-p.normal.y * 0.5 + 0.5) * viewport_size.y

	return res

func _get_camera_projection(p_near):
	var viewport_size : Rect2 = get_viewport().get_camera_rect_size()
	var cm

	match (projection):
		PROJECTION_PERSPECTIVE: 
			cm.set_perspective(fov, viewport_size.aspect(), p_near, _far, keep_aspect == KEEP_WIDTH)
		PROJECTION_ORTHOGONAL: 
			cm.set_orthogonal(size, viewport_size.aspect(), p_near, _far, keep_aspect == KEEP_WIDTH)
		PROJECTION_FRUSTUM: 
			cm.set_frustum(size, viewport_size.aspect(), frustum_offset, p_near, _far, keep_aspect == KEEP_WIDTH)

	return cm
```
I'm not sure what `p(get_camera_transform().xform_inv(p_pos), 1.0)` is mean to be and I can't find a substitute for `viewport_size.aspect()`. For my purposes, I'm using a subviewport (viewportcontainer in 3.6) so maybe I could adjust that to match, but I'd still appreciate if anyone could convert this into actual working GDScript as I'm not sure if this is going to work very well, even if I can figure out how to finish it.

elvisish commented:
> This is what AI spewed out when i let it to rewrite my C# script to GDScript. So you can start there. You should be able to call statically Workaround.unproject_position(camera3d_node, p_pos_vector3) method. [Workaround-gd.txt](https://github.com/user-attachments/files/22290190/Workaround-gd.txt)

I keep having trouble figuring out what `Projection` type should be in 3.6:
<img width="1145" height="166" alt="image" src="https://github.com/user-attachments/assets/a6fed63f-d895-495d-b389-e24ada45c33e" />

elvisish commented:
> This is what AI spewed out when i let it to rewrite my C# script to GDScript. So you can start there. You should be able to call statically Workaround.unproject_position(camera3d_node, p_pos_vector3) method. [Workaround-gd.txt](https://github.com/user-attachments/files/22290190/Workaround-gd.txt)

I'm having trouble figuring out how to use the `Projection` type:
<img width="1145" height="166" alt="image" src="https://github.com/user-attachments/assets/d5e65c3d-10f8-4e90-b03b-a46e5287f3b7" />

l0g0sys commented:
No. You cannot replace any type. All types are correct in that GD script. _get_camera_projection method must return Projection.

l0g0sys commented:
> > No. You cannot replace any type. All types are correct in that GD script. _get_camera_projection method must return Projection.
> 
> I don't think it was called by a different name in 3.x, so it doesn't exist until 4.x? https://docs.godotengine.org/en/3.6/classes/class_projection.html

Ah, there is one but it's not public and it's called CameraMatrix in core/math source folder. So it will be more complicated to backport to 3.6.

```
bool Camera::safe_unproject_position(const Vector3 &p_pos, Point2 &r_result) const {
	ERR_FAIL_COND_V_MSG(!is_inside_tree(), false, "Camera is not inside scene.");

	Size2 viewport_size = get_viewport()->get_visible_rect().size;

	CameraMatrix cm;

	if (mode == PROJECTION_ORTHOGONAL) {
		cm.set_orthogonal(size, viewport_size.aspect(), near, far, keep_aspect == KEEP_WIDTH);
	} else {
		cm.set_perspective(fov, viewport_size.aspect(), near, far, keep_aspect == KEEP_WIDTH);
	}
	...
```

So you will need to replicate this CameraMatrix with some methods and maybe operators to make it work.
Maybe try asking in official Godot discord if anyone did that already?
You might not be only one to need this workaround for 3.x ;)

elvisish commented:
> > > No. You cannot replace any type. All types are correct in that GD script. _get_camera_projection method must return Projection.
> 
> > 
> 
> > I don't think it was called by a different name in 3.x, so it doesn't exist until 4.x? https://docs.godotengine.org/en/3.6/classes/class_projection.html
> 
> 
> 
> Ah, there is one but it's not public and it's called CameraMatrix in core/math source folder. So it will be more complicated to backport to 3.6.
> 
> 
> 
> ```
> 
> bool Camera::safe_unproject_position(const Vector3 &p_pos, Point2 &r_result) const {
> 
> 	ERR_FAIL_COND_V_MSG(!is_inside_tree(), false, "Camera is not inside scene.");
> 
> 
> 
> 	Size2 viewport_size = get_viewport()->get_visible_rect().size;
> 
> 
> 
> 	CameraMatrix cm;
> 
> 
> 
> 	if (mode == PROJECTION_ORTHOGONAL) {
> 
> 		cm.set_orthogonal(size, viewport_size.aspect(), near, far, keep_aspect == KEEP_WIDTH);
> 
> 	} else {
> 
> 		cm.set_perspective(fov, viewport_size.aspect(), near, far, keep_aspect == KEEP_WIDTH);
> 
> 	}
> 
> 	...
> 
> ```
> 
> 
> 
> So you will need to replicate this CameraMatrix with some methods and maybe operators to make it work.
> 
> Maybe try asking in official Godot discord if anyone did that already?
> 
> You might not be only one to need this workaround for 3.x ;)

Thanks, I have asked a little and I've even opened a job post in the unofficial server, but I don't think anyone's interested unfortunately.

It's a shame as I make sure heavy use of the frustum camera, but I can't even really fake it either with offsets or a hidden rotation spatial because it just aims so much differently to a real rotation.

Would the entire CameraMatrix need reimplementing in gdscript?

elvisish commented:
In case it's of use to anyone, I'm just using this to get a close approximation:
```swift
var center = viewport.size * 0.5
var fake_frust = viewport.size.y * 108.7866108786611 * -1
	var to_shear = cam.project_ray_origin(center) + cam.project_ray_normal(center + Vector2(0,cam.get_frustum_offset().y * fake_frust)) * ray_length
	```
I don't know why `108.7866108786611` works, but it does, the magic number can probably be adjusted if it doesn't work with a particular set-up.

bentles commented:
was also having issues with the frustum projection mode with an offset. The offset lets my game look a lot better but I was sad to see ray tracing was completely broken after switching... this change seems to fix that for me <3 what's the process for getting the change merged?

Flarkk commented:
Rebased
